sync-core/src/main/java/com/cloudant/sync/query/QueryResult.java

```diff
@@ -44,23 +42,19 @@ public QueryResult(List<String> docIds,
                        long skip,
                        long limit,
                        UnindexedMatcher matcher) {
-        this.docIds = docIds;
+        this.docIds = matcher == null ? processDocIds(docIds, skip, limit) :
+                                        processDocIds(docIds, datastore, skip, limit, matcher);
```

Comment:
This processing needs to happen during the iteration process to avoid false positives/negatives as the contents of the document store could change between this point and the iterator.